### PR TITLE
ARROW-9641: [C++][Gandiva] Implement round() for floating point and double floating point numbers

### PR DESCRIPTION
**Pull request by sgnkc**

### CODE REVIEW

File: cpp/src/gandiva/precompiled/extended_math_ops_test.cc [hunk 1]
```diff
@@ -87,6 +87,20 @@ TEST(TestExtendedMathOps, TestLogWithBase) {
...
 }
 
+TEST(TestExtendedMathOps, TestRoundDecimal) {
+  EXPECT_FLOAT_EQ(round_float32_int32(1234.789, 2), 1234.79);
+  EXPECT_FLOAT_EQ(round_float32_int32(1234.12345, -3), 1000);
+  EXPECT_FLOAT_EQ(round_float32_int32(-1234.4567, 3), -1234.457);
+  EXPECT_FLOAT_EQ(round_float32_int32(-1234.4567, -3), -1000);
+  EXPECT_FLOAT_EQ(round_float32_int32(1234.4567, 0), 1234);
+
+  VerifyFuzzyEquals(round_float64_int32(1234.789, 2), 1234.79);
```
Comment (kiszk) [inline cpp/src/gandiva/precompiled/extended_math_ops_test.cc:97]: What happens when we convert (double)INT_MAX+1 or (double)INT_MIN-1 int32?

Comment (sagnikc-dremio) [reply cpp/src/gandiva/precompiled/extended_math_ops_test.cc:97]: I am sorry, I did not fully understand your concern. Are you suggesting that some unit tests be added for `round_float64_int32()` where the first argument is INT_MAX+1 or INT_MIN-1?

Comment (kiszk) [reply cpp/src/gandiva/precompiled/extended_math_ops_test.cc:97]: Sure, there are examples. I do not know the results in `int32`.

```
VerifyFuzzyEquals(round_float64_int32((double)INT_MAX+1, 0), ???);
VerifyFuzzyEquals(round_float64_int32((double)INT_MIN-1, 0), ???);

```

Comment (sagnikc-dremio) [reply cpp/src/gandiva/precompiled/extended_math_ops_test.cc:97]: Added the unit tests.

File: cpp/src/gandiva/precompiled/extended_math_ops.cc [hunk 2]
```diff
@@ -111,6 +111,20 @@ LOG_WITH_BASE(float64, float64, float64)
 
 POWER(float64, float64, float64)
 
+// round
+#define ROUND_DECIMAL(TYPE)                                                 \
+  FORCE_INLINE                                                              \
+  gdv_##TYPE round_##TYPE##_int32(gdv_##TYPE number, gdv_int32 precision) { \
```
Comment (praveenbingo) [inline cpp/src/gandiva/precompiled/extended_math_ops.cc:117]: should this be called output scale?

Comment (sagnikc-dremio) [reply cpp/src/gandiva/precompiled/extended_math_ops.cc]: Yes, made the change.

**Review by praveenbingo — APPROVED** [review]

File: cpp/src/gandiva/precompiled/types.h [hunk 3]
```diff
@@ -132,6 +132,10 @@ gdv_int64 div_int64_int64(gdv_int64 context, gdv_int64 in1, gdv_int64 in2);
 gdv_float32 div_float32_float32(gdv_int64 context, gdv_float32 in1, gdv_float32 in2);
 gdv_float64 div_float64_float64(gdv_int64 context, gdv_float64 in1, gdv_float64 in2);
 
+gdv_float32 round_float32_int32(gdv_float32 number, gdv_int32 precision);
```
Comment (praveenbingo) [inline cpp/src/gandiva/precompiled/types.h:135]: looks like precision is misnamed..can you please rename this too

Comment (sagnikc-dremio) [reply cpp/src/gandiva/precompiled/types.h:135]: Oh right sure. Missed that.

**Pull request closed without merging**